frigate/edgetpu.py

```diff
@@ -177,6 +179,14 @@ def detect(self, tensor_input, threshold=.4):
         self.np_shm[:] = tensor_input[:]
         self.event.clear()
         self.detection_queue.put(self.name)
+
+        if self.config["saveTensorInputs"]:
+            root_path = self.config['saveTensorPath']
+            os.makedirs(root_path, exist_ok=True)
+
+            file_path = os.path.join(root_path, f"{self.name}.{datetime.datetime.now().timestamp()}.jpg")
+            cv2.imwrite(file_path, tensor_input[0])
```

Comment:
This code executes synchronously and the function will not return until the image is written to disk. This is in the critical path for this camera's detection pipeline and will add time to each detection call. Frigate often calls this function several times per frame.

Reply:
@blakeblackshear That's not an accurate assessment: 

On L179:
`self.detection_queue.put(self.name)`

places the frame tensor in the detection queue of the "LocalObjectDetector" process.

on L180:
`result = self.event.wait(timeout=10.0)`

Waits until the process is completed on the "LocalObjectDetector". 

The proposed code executes **in parallel** with the inference running on the LocalObjectDetector, when normally would be sleeping on `self.event.wait`, which internally calls `self._cond.wait(timeout)` on a sempahore in Multiprocessing.

So unless the inference time is much faster the time it takes `imwrite` to return, or there aren't enough spare threads on the system (cpu constrained for other reasons) writing the image does not affect the critical path. On my system, I've benchmarked imwrite, and it takes 1-4ms for it to return (remember, it is not writing to disk but the pagecache), far below the time it takes for an inference to run. It doesn't matter how many times frigate calls this function, it **does not normally add time to the detection path**

In fact, if you want to get pedantic, it seems that writing the file to disk counterintuitively **improves** the  detection speed occasionally, probably because the event becomes ready when the wait call is encountered, so the OS doesn't try to context the switch out.

Reply:
Well now I feel dumb. You are totally right. This little window of time is a great opportunity to do extra work. My apologies.

Reply:
No big deal! Besides, I think you've managed to convince me that this is stuff that's better to do offline than in real time. Maybe we can run saved clips through time same motion filtering that frigate is already doing.

Reply:
that would be great to recognize specific cars, be able to run Openalpr on those only, and then trigger actions.
Bottom line, being able to train to recognize specific cars instead of "car" would be good.